mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java

```diff
@@ -55,6 +55,12 @@ public static void load(Context context, EnvoyEngine envoyEngine) {
     }
   }
 
+  /**
+   * Sets the {@link AndroidNetworkMonitor} singleton instance to null, so that it can be recreated
+   * when a new EnvoyEngine is created.
+   */
+  public static void shutdown() { instance = null; }
```

Comment:
For multi-engine support, the existing `AndroidNetworkMonitor` doesn't work and will need to be updated anyway.

Removing the test will reduce the coverage. I updated the code to use `@VisibleForTesting` annotation, which indicates that the method is only for testing. This is similar to https://chromium.googlesource.com/chromium/src/+/HEAD/docs/patterns/fortesting-methods.md

Reply:
That certainly seems like an expedient choice in the short term. I do think multi-engine will be valuable in the general case. But maybe we don't have the bandwidth to worry about it now?

Reply:
To add on further (not really related to multi-engine) Before the PR this would introduce a bug.
1. Start engine 1
2. Terminate engine 1
3. Start engine 2 -- bug! It's still holding on engine 1 instance

Reply:
> Yes because the singleton is per EnvoyEngine

*shudder* This seems fundamentally problematic, right? Seems like a pretty clear anti-pattern. (And is clearly a long-standing issue, right?)

Reply:
> > Yes because the singleton is per EnvoyEngine
> 
> _shudder_ This seems fundamentally problematic, right? Seems like a pretty clear anti-pattern. (And is clearly a long-standing issue, right?)

Yup. I could fix it here, too but given this PR is about the issue around `terminate` and supporting multi-engine is a low priority, I basically did a bare minimum to keep the existing tests happy while keeping the coverage 😄

Reply:
Yeah, this is definitely problematic.  Given it's a long standing issue, I think we should land this PR since it's a strict improvement, and address the AndroidNetworkMonitor singleton issue in a follow up PR.

Reply:
SGTM.  I'm out tomorrow so I'll just take another look if this hasn't landed by Monday.